# Vehicle Sales Data: Exploratory Data Analysis (EDA)

### **Introduction**
This notebook provides a comprehensive exploratory analysis of a dataset containing vehicle sales data. The aim is to uncover trends, identify patterns, and prepare the data for potential applications such as machine learning models or interactive web applications. Key tasks include:

- **Data Preprocessing**:
  - Removing duplicates.
  - Handling missing values using logical and contextual imputations.
  - Summarizing the dataset to identify key statistics and characteristics.

- **Data Visualization**:
  - Creating visualizations to explore price distributions, trends in model years, and the impact of vehicle condition and fuel type.

- **Key Objectives**:
  - Understand the data structure and quality.
  - Derive actionable insights to assist in strategic decision-making.
  - Prepare the dataset for further analysis or application development.

With this approach, the notebook combines data cleaning, exploration, and visualization to deliver a thorough understanding of the dataset.


In [7]:
import pandas as pd
import plotly.express as px



## Step 1: Load the Dataset

### **Code Explanation**
- The dataset `vehicles_us.csv` is loaded into a Pandas DataFrame.
- The shape of the dataset is printed using `df.shape`, which shows the total number of rows and columns in the dataset.
- The first 51 rows of the dataset are displayed using `df.head(51)` to give an overview of the data.


In [19]:
print("Step 1: Load the Dataset")
df = pd.read_csv('vehicles_us.csv')
print(f"Initial Shape: {df.shape}")
print(df.head(51))

Step 1: Load the Dataset
Initial Shape: (51525, 13)
    price  model_year                              model  condition  \
0    9400      2011.0                             bmw x5       good   
1   25500         NaN                         ford f-150       good   
2    5500      2013.0                     hyundai sonata   like new   
3    1500      2003.0                         ford f-150       fair   
4   14900      2017.0                       chrysler 200  excellent   
5   14990      2014.0                       chrysler 300  excellent   
6   12990      2015.0                       toyota camry  excellent   
7   15990      2013.0                        honda pilot  excellent   
8   11500      2012.0                        kia sorento  excellent   
9    9200      2008.0                        honda pilot  excellent   
10  19500      2011.0           chevrolet silverado 1500  excellent   
11   8990      2012.0                       honda accord  excellent   
12  18990      2012.0    

1. **Dataset Shape**  
   - The dataset contains **515,525 rows** and **13 columns**.

2. **Columns Present**
   - `price`: The price of the vehicle (numerical).
   - `model_year`: The year the vehicle model was manufactured (numerical, with some missing values).
   - `model`: The name of the vehicle model (categorical).
   - `condition`: The condition of the vehicle (categorical: `excellent`, `good`, `like new`, etc.).
   - Additional columns such as `fuel`, `odometer`, `paint_color`, etc.

3. **Observations**
   - **Missing Values:** Some rows have missing values in the `model_year` column (e.g., rows with `NaN` values).
   - **Data Types:** The dataset contains a mix of numerical and categorical columns.
   - **Price Range:** The `price` column has a wide range of values, potentially requiring further analysis or cleaning (e.g., handling outliers).

# Step 2: Check for Duplicates

### **Code Explanation**
- The code checks for duplicate rows in the dataset and removes them if any are found.
- **`df.duplicated().sum()`**: Identifies the total number of duplicate rows in the dataset.
- **`print(f"Number of duplicate rows: {duplicates_count}")`**: Displays the count of duplicate rows found.
- **`df.drop_duplicates()`**: Removes the duplicate rows from the dataset.


In [20]:
print("\nStep 2: Check for Duplicates")
duplicates_count = df.duplicated().sum()
print(f"Number of duplicate rows: {duplicates_count}")
df = df.drop_duplicates()


Step 2: Check for Duplicates
Number of duplicate rows: 0


### **Output Findings**
1. **Number of Duplicate Rows**
   - The code found **0 duplicate rows** in the dataset.

2. **Dataset Integrity**
   - Since no duplicates were present, the dataset remains unchanged.


# Step 3: Handle Missing Values

### **Code Explanation**
- The code identifies and summarizes the missing values in each column of the dataset.
- **`df.isnull().sum()`**: Calculates the total number of missing values in each column of the dataset.
- **`print(missing_summary)`**: Displays the missing values summary for better understanding of the data's quality.


In [21]:
print("\nStep 3: Handle Missing Values")
missing_summary = df.isnull().sum()
print("Missing Values Summary:")
print(missing_summary)


Step 3: Handle Missing Values
Missing Values Summary:
price               0
model_year       3619
model               0
condition           0
cylinders        5260
fuel                0
odometer         7892
transmission        0
type                0
paint_color      9267
is_4wd          25953
date_posted         0
days_listed         0
dtype: int64


### **Output Findings**
1. **Missing Values Summary**
   - The output highlights the columns with missing values and their respective counts:
     - `model_year`: 3,619 missing values.
     - `cylinders`: 5,260 missing values.
     - `odometer`: 7,892 missing values.
     - `paint_color`: 9,267 missing values.
     - `is_4wd`: 25,953 missing values.
   - Other columns such as `price`, `model`, `condition`, and `fuel` have no missing values.

2. **Dataset Quality**
   - The missing values indicate that certain features are incomplete and require imputation or handling to ensure proper analysis.


In [22]:
if 'is_4wd' in df.columns:
    df['is_4wd'] = df['is_4wd'].fillna(0).astype(int)
if 'paint_color' in df.columns:
    df['paint_color'] = df['paint_color'].fillna('unknown')

# Handling Missing Values for Specific Columns

### **Code Explanation**
- This snippet focuses on handling missing values in the `is_4wd` and `paint_color` columns.
- **`if 'is_4wd' in df.columns:`**:
  - Checks if the column `is_4wd` exists in the dataset.
  - Missing values in this column are filled with `0` (assuming vehicles without 4WD are represented by 0).
  - The column is then converted to an integer type using `.astype(int)`.
- **`if 'paint_color' in df.columns:`**:
  - Checks if the column `paint_color` exists in the dataset.
  - Missing values in this column are replaced with the placeholder `"unknown"`, indicating that the color is not specified.


### **Output Findings**
1. **Column: `is_4wd`**
   - All missing values are replaced with `0`.
   - The column is standardized as an integer type, ensuring consistent data representation.

2. **Column: `paint_color`**
   - Missing values are replaced with `"unknown"`, making it easier to categorize and analyze.


### **Benefits**
- Ensures the dataset is complete and consistent for these columns.
- Avoids errors caused by null values during analysis or modeling.
- Improves interpretability by clearly distinguishing between missing and meaningful data.


In [23]:
if 'model_year' in df.columns:
    df['model_year'] = df['model_year'].fillna(
        df.groupby(['model'])['model_year'].transform('median')
    ).astype(int)

# Imputing Missing Values in `model_year`

### **Code Explanation**
- **`if 'model_year' in df.columns:`**
  - Checks if the `model_year` column exists in the dataset before attempting to process it.
- **Handling Missing Values:**
  - Missing values in `model_year` are filled using the median `model_year` grouped by the corresponding `model`.
  - This is achieved through:
    - **`df.groupby(['model'])['model_year'].transform('median')`:**
      - Groups the dataset by `model`.
      - Calculates the median `model_year` for each group.
  - The missing values in `model_year` are then filled with the respective median value of their `model`.
- **Converting to Integer Type:**
  - After filling the missing values, the column is converted to integers using `.astype(int)` for consistency.


### **Output Findings**
1. **Column: `model_year`**
   - Missing values are replaced with the median `model_year` of the corresponding `model`.
   - This ensures that the imputation leverages meaningful relationships within the data.
   - The column values are standardized as integers.


### **Benefits**
- Uses a contextual imputation strategy based on related characteristics (`model`), improving data integrity.
- Reduces the likelihood of bias that may arise from filling missing values with a single global value.
- Ensures consistency and prepares the data for further analysis or modeling.


In [24]:
if 'odometer' in df.columns:
    df['odometer'] = df['odometer'].fillna(
        df.groupby(['model'])['odometer'].transform('median')
    )


# Imputing Missing Values in `odometer`

### **Code Explanation**
- **`if 'odometer' in df.columns:`**
  - Ensures that the `odometer` column is present in the dataset before performing operations.
- **Handling Missing Values:**
  - The missing values in the `odometer` column are imputed using the median `odometer` value grouped by `model`.
  - **Grouping by `model`:**
    - **`df.groupby(['model'])['odometer'].transform('median')`:**
      - Groups the dataset by the `model` column.
      - Computes the median `odometer` value for each model group.
  - **Filling Missing Values:**
    - Missing `odometer` values are filled with the corresponding median value for their model group.
- This approach ensures that the imputed values are contextually meaningful and aligned with the characteristics of each `model`.


### **Output Findings**
1. **Column: `odometer`**
   - Missing values are replaced with the median `odometer` of the respective `model`.
   - This improves the reliability of the data by avoiding arbitrary global imputations.
2. Ensures that the `odometer` column reflects realistic values derived from its associated `model`.


### **Benefits**
- The imputation is specific to each model, capturing inherent variations in odometer readings between different vehicle models.
- Enhances the dataset's integrity, making it better suited for subsequent analyses, such as price prediction or trend discovery.


In [25]:
if 'cylinders' in df.columns:
    df['cylinders'] = df['cylinders'].fillna(
        df.groupby(['model'])['cylinders'].transform('median')
    )

# Imputing Missing Values in `cylinders`

### **Code Explanation**
- **`if 'cylinders' in df.columns:`**
  - Ensures that the `cylinders` column is present in the dataset before performing any operations.
- **Handling Missing Values:**
  - The missing values in the `cylinders` column are imputed using the median `cylinders` value grouped by `model`.
  - **Grouping by `model`:**
    - **`df.groupby(['model'])['cylinders'].transform('median')`:**
      - Groups the dataset by the `model` column.
      - Computes the median number of cylinders for each model group.
  - **Filling Missing Values:**
    - Missing `cylinders` values are replaced with the corresponding median number of cylinders for their model group.
- This targeted imputation ensures that the filled values are consistent with the mechanical specifications of each vehicle model.


### **Output Findings**
1. **Column: `cylinders`**
   - Missing values are filled with the median `cylinders` value for the respective `model`.
   - The updated `cylinders` data becomes more representative of real-world vehicle configurations.
2. Eliminates the risk of assigning arbitrary values across the dataset, maintaining contextual relevance.


### **Benefits**
- Captures engine-specific variations in the number of cylinders between different models.
- Enhances the dataset's reliability, especially for analyses involving performance or engine-related attributes.


# Step 4: Dataset Summary

### **Code Explanation**
- **Purpose:**
  - The code provides a statistical summary of the dataset using the `describe()` function.
  - This summary includes key metrics for numerical columns such as `price`, `model_year`, `cylinders`, `odometer`, `is_4wd`, and `days_listed`.

- **`print(df.describe())`:**
  - Outputs descriptive statistics:
    - **Count:** Number of non-null entries in each column.
    - **Mean:** Average value of the column.
    - **Standard Deviation (std):** Measure of data spread or dispersion.
    - **Min and Max:** Minimum and maximum values in the column.
    - **Quartiles (25%, 50%, 75%):** Percentile-based summary of data distribution.


In [26]:
print("\nStep 4: Dataset Summary")
print(df.describe())


Step 4: Dataset Summary
               price    model_year     cylinders       odometer        is_4wd  \
count   51525.000000  51525.000000  51525.000000   51484.000000  51525.000000   
mean    12132.464920   2009.793557      6.121067  115338.061971      0.496303   
std     10040.803015      6.099381      1.657457   60638.631682      0.499991   
min         1.000000   1908.000000      3.000000       0.000000      0.000000   
25%      5000.000000   2007.000000      4.000000   77000.000000      0.000000   
50%      9000.000000   2011.000000      6.000000  115376.000000      0.000000   
75%     16839.000000   2014.000000      8.000000  148177.000000      1.000000   
max    375000.000000   2019.000000     12.000000  990000.000000      1.000000   

       days_listed  
count  51525.00000  
mean      39.55476  
std       28.20427  
min        0.00000  
25%       19.00000  
50%       33.00000  
75%       53.00000  
max      271.00000  


### **Output Findings**
1. **`price`:**
   - Average price: **\$12,123.46**
   - Minimum price: **\$1**
   - Maximum price: **\$375,000**
   - Majority of vehicles are priced below \$16,839 (75th percentile), with some high-value outliers.

2. **`model_year`:**
   - Most vehicles are manufactured around **2010** (mean model year: 2009.79).
   - Oldest vehicle: **1908**.
   - Newest vehicle: **2019**.

3. **`cylinders`:**
   - Average number of cylinders: **6.12**.
   - Range: **3 to 12 cylinders**, reflecting diverse engine configurations.

4. **`odometer`:**
   - Average mileage: **115,338 miles**.
   - Range: **0 to 990,000 miles**.
   - Data includes a mix of low-mileage and high-mileage vehicles, possibly impacting resale values.

5. **`is_4wd`:**
   - Binary column indicating 4-wheel drive capability.
   - About 49.63% of vehicles have 4WD (`mean` ~ 0.496).

6. **`days_listed`:**
   - Average days listed for sale: **39.55 days**.
   - Minimum: **0 days** (just listed or removed immediately).
   - Maximum: **271 days** (indicating vehicles that take longer to sell).


### **Insights**
- **Price:**
  - Vehicles are generally affordable, with most below \$20,000.
  - High-value vehicles likely represent luxury or rare models.

- **Model Year:**
  - Majority of vehicles are relatively recent (post-2007).
  - Distribution indicates a healthy mix of older and newer models.

- **Cylinders:**
  - Data reflects a variety of engine configurations, suitable for different vehicle types (e.g., compact cars vs. trucks).

- **Odometer:**
  - Wide range suggests inclusion of both used and heavily-used vehicles.
  - Low-mileage vehicles may align with higher prices.


# Visualization: Price Distribution

### **Code Explanation**
- **Purpose:**
  - The code generates a histogram to visualize the distribution of vehicle prices.
  - Histograms are useful to identify pricing trends, clusters, and outliers.

- **Key Components:**
  - **`px.histogram()`**:
    - `df`: The dataset being analyzed.
    - `x='price'`: Specifies that the `price` column is used to create the histogram.
    - `nbins=60`: Sets the number of bins to 60, dividing the price range into equal intervals.
    - `title="Distribution of Prices"`: Sets the title for the plot.
  - **`update_layout()`**:
    - `xaxis_title="Price ($)"`: Adds a label to the x-axis to indicate prices in dollars.
    - `yaxis_title="Count"`: Adds a label to the y-axis to represent the number of vehicles in each price range.
  - **`show()`**: Displays the final plot.


In [27]:
print("\nVisualization: Price Distribution")
price_histogram = px.histogram(
    df,
    x='price',
    nbins=60,
    title="Distribution of Prices"
)
price_histogram.update_layout(xaxis_title="Price ($)", yaxis_title="Count")
price_histogram.show()


Visualization: Price Distribution


### **Output Analysis**
- The histogram shows the **count of vehicles** for different price ranges.
- Key observations:
  1. **Most vehicles are priced below \$20,000**.
  2. The count decreases significantly for vehicles priced above \$50,000.
  3. A few vehicles are extremely expensive, reaching up to \$375,000 (likely luxury or rare models).


### **Insights**
- **Affordable Vehicles Dominate**:
  - The dataset includes a high number of vehicles within the lower price range, making it more relevant to budget-conscious buyers.
- **Outliers**:
  - Vehicles priced above \$100,000 are clear outliers and may represent luxury brands or specialized models.


# Visualization: Price vs Model Year

### **Code Explanation**
- **Purpose:**
  - The code creates a scatter plot to examine how vehicle prices vary with the model year.
  - The `condition` of the vehicle is used to color-code the points, providing additional insight.

- **Key Components:**
  - **`px.scatter()`**:
    - `df`: Specifies the dataset.
    - `x='model_year'`: Assigns the `model_year` column to the x-axis.
    - `y='price'`: Assigns the `price` column to the y-axis.
    - `color='condition'`: Colors the points based on the `condition` of the vehicle, allowing for condition-based analysis.
    - `title="Price vs Model Year"`: Sets the title of the scatter plot.
    - `labels`: Customizes axis labels for clarity (`Model Year` and `Price ($)`).
  - **`show()`**:
    - Displays the scatter plot.

In [28]:
print("\nVisualization: Price vs Model Year")
scatter_plot = px.scatter(
    df,
    x='model_year',
    y='price',
    color='condition',
    title="Price vs Model Year",
    labels={'model_year': 'Model Year', 'price': 'Price ($)'},
)
scatter_plot.show()


Visualization: Price vs Model Year


### **Output Analysis**
- The scatter plot reveals the **relationship between vehicle prices and model years**.
- Key observations:
  1. **Older vehicles** (pre-2000) generally have lower prices, as expected due to depreciation.
  2. **Newer vehicles** (post-2010) exhibit a wide price range, with some outliers exceeding \$300,000.
  3. The **condition** of vehicles plays a significant role in pricing:
     - **"New" and "Excellent" vehicles** are often priced higher.
     - **"Salvage" vehicles** tend to be clustered at lower price points.


### **Insights**
- **Trend with Age**:
  - The plot confirms that vehicle prices increase with the recency of the model year, reflecting technological advancements, newer designs, and less wear-and-tear.
- **Condition-Based Pricing**:
  - Vehicles in better condition command a premium, with "New" vehicles being the most expensive.
- **Outliers**:
  - A few high-priced vehicles in earlier model years might indicate classic or collector cars.


# Visualization: Count of Cars by Fuel Type

### **Code Explanation**
- **Purpose:**
  - This code generates a bar chart to display the distribution of cars based on their fuel types.
  - It helps to identify which fuel types dominate the dataset and analyze their relative proportions.

- **Key Components:**
  - **`value_counts()`**:
    - Counts the occurrences of each fuel type in the `fuel` column.
    - Creates a summary of fuel types and their respective counts.
  - **`reset_index()`**:
    - Resets the index of the resulting summary to make it a DataFrame.
    - Facilitates renaming columns for clarity.
  - **`px.bar()`**:
    - `x='fuel'`: Sets the fuel types as the x-axis.
    - `y='count'`: Sets the count of each fuel type as the y-axis.
    - `title="Count of Cars by Fuel Type"`: Sets the title of the bar chart.
    - `labels`: Customizes the axis labels (`Fuel Type` and `Number of Cars`).
    - `color='fuel'`: Colors each bar according to the fuel type for visual differentiation.
  - **`show()`**:
    - Displays the bar chart.


In [29]:
print("\nVisualization: Count of Cars by Fuel Type")
fuel_counts = df['fuel'].value_counts().reset_index()
fuel_counts.columns = ['fuel', 'count']

bar_chart = px.bar(
    fuel_counts,
    x='fuel',
    y='count',
    title="Count of Cars by Fuel Type",
    labels={'fuel': 'Fuel Type', 'count': 'Number of Cars'},
    color='fuel'
)
bar_chart.show()



Visualization: Count of Cars by Fuel Type


### **Output Analysis**
- The bar chart reveals the distribution of cars across different fuel types:
  - **Gasoline (gas)**:
    - Dominates the dataset with over 40,000 vehicles.
    - Reflects the popularity and prevalence of gasoline-powered vehicles in the market.
  - **Diesel**:
    - Represents a smaller but significant share, often associated with trucks and heavy-duty vehicles.
  - **Hybrid, Electric, and Other**:
    - These categories have minimal representation, indicating their limited adoption compared to traditional fuel types.


### **Insights**
1. **Gas-Powered Vehicles**:
   - The dominance of gasoline suggests it remains the preferred choice for most vehicles, likely due to cost, availability, and infrastructure.
2. **Emerging Fuel Types**:
   - Hybrid and electric vehicles have a growing but minimal share, reflecting their potential for future growth as the industry shifts toward greener technologies.
3. **Diesel Market**:
   - Diesel vehicles cater to niche markets, often involving heavy-duty applications and specific use cases.


In [30]:
print("\nSaving cleaned dataset...")
df.to_csv('vehicles_us_cleaned.csv', index=False)
print("Cleaned dataset saved as 'vehicles_us_cleaned.csv'")


Saving cleaned dataset...
Cleaned dataset saved as 'vehicles_us_cleaned.csv'


### **Conclusion**
This exploratory data analysis of vehicle sales data yielded several important findings and actionable insights:

1. **Data Cleaning**:
   - Removed duplicate rows, ensuring data consistency.
   - Addressed missing values for critical columns such as `model_year`, `odometer`, and `cylinders` using contextual imputations (e.g., grouping by model and calculating medians).
   - Ensured categorical variables like `is_4wd` and `paint_color` were logically imputed with default values.

2. **Key Insights**:
   - **Price Trends**:
     - Most vehicles are priced below $20,000, with a few outliers priced above $300,000.
   - **Vehicle Age and Price**:
     - Newer vehicles (post-2010) tend to have higher prices, as seen in the scatter plot of `Price vs Model Year`.
   - **Fuel Type Dominance**:
     - Gasoline-powered vehicles dominate the dataset, followed by diesel vehicles, with minimal representation of hybrids and electric vehicles.
   - **Condition and Price**:
     - "New" vehicles command the highest prices, while "Fair" and "Salvage" vehicles are priced lowest, as observed in the box plot analysis.

3. **Next Steps**:
   - Use the cleaned and processed dataset to develop advanced applications, such as predictive models or interactive dashboards.
   - Incorporate the visualizations and findings into user-friendly tools to aid decision-making in vehicle sales and marketing strategies.

By combining robust data preprocessing techniques with insightful visualizations, this analysis sets the foundation for actionable outcomes, from business strategy to data-driven application development.
